### ISSUE TITLE
Yandex: support deduplication on upload

### ISSUE THREAD

**Issue by Evengard** [rclone/rclone#6032]
#### What is your current rclone version (output from `rclone version`)?
1.57.0

#### What problem are you are trying to solve?
Yandex supports deduplication on file upload - you need to pass the Md5 and Sha256 hashes and the file size to the PUT endpoint, and if a file like that exists anywhere on Yandex - the endpoint answers with "201 File Exists" and the upload is finished, as the already existing file is made available.

#### How do you think rclone should be changed to solve that?
Rclone should probably implement it! =) Now, this involves calculating a SHA256 hash on each file upload, so maybe the feature should be placed behind a parameter.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Can you link to the API docs?

Maybe you'd like to implement this? I don't think it would be too hard?

**Comment by Evengard** [comment]
It's kind of in russian, but here is the doc: https://yandex.ru/dev/disk/doc/dg/reference/put.html
The example is mostly self-explanatory.

I can try, but I don't know how to add new setting parameters (and if needed at all, actually).

**Comment by ncw** [comment]
Luckily google translate works really well on that page!

> If a file with the same size, MD5 sum, and SHA-256 hash already exists in the store, the service returns a 201 Created.
> 
> File data should be sent in the request headers: size in the header Content-Length, MD5 sum in the header Etag, and SHA hash in the header Sha256.

So you send the md5 sum as the Etag, and the Sha256 header and look for a 201 created response.

This also implies that we are using expect/continue which I think we are but may need to be enabled.

> Now, this involves calculating a SHA256 hash on each file upload, so maybe the feature should be placed behind a parameter.

sha256 do cost a bit to calculate, but compared to the upload cost it is small

```
$ truncate -s 1GB 1GB
$ time rclone md5sum 1GB 
2022/03/13 14:59:55 NOTICE: OSL_PIPE_1000_SingleOfficeIPC_41ab247d8417b7558eafae039408995: Can't transfer non file/directory
2022/03/13 14:59:55 NOTICE: dbus-f04udQjV82: Can't transfer non file/directory
2022/03/13 14:59:55 NOTICE: .#query.sql: Can't follow symlink without -L/--copy-links
e37115d4da0e187130ab645dee4f14ed  1GB

real	0m2.091s
user	0m1.874s
sys	0m0.571s
$ time rclone hashsum sha256 1GB 
2022/03/13 15:00:13 NOTICE: OSL_PIPE_1000_SingleOfficeIPC_41ab247d8417b7558eafae039408995: Can't transfer non file/directory
2022/03/13 15:00:13 NOTICE: dbus-f04udQjV82: Can't transfer non file/directory
2022/03/13 15:00:13 NOTICE: .#query.sql: Can't follow symlink without -L/--copy-links
bc17f06f9d9b5f6f79ca189a1772b1a3a38d6e40c45bec50f9c4f28144efddca  1GB

real	0m3.464s
user	0m3.357s
sys	0m0.512s
$ 
```

Only the local backend will have sha256 and md5sum available (according to [the overview](https://rclone.org/overview/)) so this will be a local backend only feature.

I had a quick go at this here

[v1.58.0-beta.6030.7b3caf8d1.fix-6032-yandex-upload](https://beta.rclone.org/branch/fix-6032-yandex-upload/v1.58.0-beta.6030.7b3caf8d1.fix-6032-yandex-upload/) on branch [fix-6032-yandex-upload](https://github.com/rclone/rclone/tree/fix-6032-yandex-upload) (uploaded in 15-30 mins)

However I wasn't convinced it was working, or at least I don't know if any file I tried was available for deduplication.

Have a go and tell me what you think. You might want to set `-expect-continue-timeout 5s` and to see what rclone is sending use `-vv --dump headers`.

**Comment by Evengard** [comment]
Sorry for the late test, I just got the time to test it out.
That didn't work... And... After reading their docs more carefully... we got a bigger problem here. 
Welp.
See, Yandex actually have TWO sets of APIs for Yandex.Disk.
One of them (actually, used in rclone) is the older REST-based one (refer https://yandex.ru/dev/disk/api/reference/upload.html)
The second one (apparently newer, and the one ACTUALLY supporting the deduplication feature) - the WebDAV-based one (basically being WebDAV but with added features on top).
Honestly, I have no ideas if they can be mixed (eg, using the WebDAV methods for actual uploads, and keeping all the rest in the REST world) - maybe yes, maybe not.

**Comment by ncw** [comment]
> That didn't work.

:-(



> And... After reading their docs more carefully... we got a bigger problem here.
> Welp.
> See, Yandex actually have TWO sets of APIs for Yandex.Disk.

Ah...



> Honestly, I have no ideas if they can be mixed (eg, using the WebDAV methods for actual uploads, and keeping all the rest in the REST world) - maybe yes, maybe not.

Probably, but the authentication might be different which would be a pain.

We've already re-written the Yandex backend once @buengese did so in 2018... Maybe it should be using the new Webdav-like interface, I'm not sure. Any views @Evengard ? Is the old REST API being deprecated?

**Comment by Evengard** [comment]
As much as I can deduce from the documentation, the REST API is NOT being deprecated - at least yet. But it seems from their blog posts that new development is mostly focused on the WebDAV API, although it seems the WebDAV API still doesn't have full feature parity with the REST API.

It seems, that the OAuth token MAY be used at the same time for both WebDAV API and REST API - although the settings for the client id of rclone may need a modification (enabling WebDAV API for it). Need more tests though.

Maybe we can keep the old REST API, and implement the new WebDAV API as a separate backend, "inheriting" it from the WebDAV implementation and adding the Yandex-specific features on top? We have already the "default WebDAV" flavor, the "OwnCloud/NextCloud" flavor, the "Sharepoint" flavor, maybe it's time for the "Yandex" flavor?

**Comment by buengese** [comment]
I'm sorry I haven't been active here in a long time. I've looked at the documentation too and I'm pretty sure @Evengard findings are correct.
The new WebDAV seems to be complete enough to mostly replace the old API and I think we should probably move to that. While not ideal it looks like it would be possible in principle to mix the old and the new api to work around missing features when using WebDAV. I'll think about it.

**Comment by devnoname120** [comment]
@Evengard @buengese This is incorrect. The REST API is [**NOT** older than the WebDAV API but newer.](https://habr.com/en/companies/yandex/articles/227377/). Where do you see that they are currently mostly focused on the WebDAV API?

Here is a post that compares the old WebDAV API with the new REST API: https://habr.com/en/companies/yandex/articles/241223/

Note that contrary to what the comparison table suggests, [free/used space property can now also be obtained from the new REST API](https://habr.com/en/companies/yandex/articles/241223/). As far as I understand the new REST API has more features than WebDAV — the only exception is the deduplication on file upload by hash which is only available in the old WebDAV API.

IMO we should keep the current implementation, and maybe add a warning if someone tries to use the Yandex WebDAV endpoint in a `webdav` remote. Adding an old WebDAV request in the current REST API `yandex` remote implementation just for the deduplication feature isn't worth it IMO given the added maintenance complexity and the risk of edge cases.

**Comment by Evengard** [comment]
Well, honestly, for me (for some reason) rclone with Yandex.Disk is broken on big files, with slow and erratic uploads, more often than not failing it.

**Comment by devnoname120** [comment]
@Evengard Could you open an issue with the logs for the file upload issues that you have with the `yandex` remote?
